47.md

```diff
@@ -0,0 +1,112 @@
+
+## Events
+
+There are two event kinds:
+- `NIP-47 request`: 23194
+- `NIP-47 response`: 23195
+
+Both the request and response events SHOULD only contain one `p` tag, containing the public key of the **wallet service** if this is a request, and the public key of the **client** if this is a response.
+
+The content is encrypted with [NIP04](https://github.com/nostr-protocol/nips/blob/master/04.md), and is a JSON object. The content depends on the kind.
+
+Request:
+```jsonc
+{
+    "cmd": "pay_invoice", // command, string
+    "data": { // data, object
+        "invoice": "lnbc50n1..." // command-related data
+    }
+}
+```
+
+Response:
+```jsonc
+{
+    "status": "ok", // status, "ok" | "error"
+    "event": "0123456789abcdef...", // event the command is in response to, string
+    "data": { // response data
+        "preimage": "0123456789abcdef..." // command-related data
+    }
+}
+```
+
+The data field SHOULD contain a `message` field with a human readable error message if the status is `error`.
+
+## Nostr Wallet Connect URI
+**client** discovers **wallet service** by scanning a QR code, handling a deeplink or pasting in a URI.
+
+The **wallet service** generates this connection URI with protocol `nostr+walletconnect:` and base path it's hex-encoded `pubkey` with the following query string parameters: 
+
+- `relay` Required. URL of the relay where the **wallet service** is connected and will be listening for events. May be more than one.
+- `secret` Required. 32-byte randomly generated hex encoded string. The **client** should use this to sign events when communicating with the **wallet service**.
+    - Authorization does not require passing keys back and forth.
+    - The user can have different keys for different applications. Keys can be revoked and created at will and have arbitrary constraints (eg. budgets).
+    - The key is harder to leak since it is not shown to the user and backed up.
+    - It improves privacy because the user's main key would not be linked to their payments.
+
+The **client** should then store this connection and use it when the user wants to perform actions like paying an invoice. Optionally it can display metadata about the connected **wallet service** from it's profile (name, image, url).
+
+### Example connection string
+```sh
+nostr+walletconnect:b889ff5b1513b641e2a139f661a661364979c5beee91842f8f0ef42ab558e9d4?relay=wss%3A%2F%2Frelay.damus.io&secret=71a8c14c1407c113601079c4302dab36460f0ccd0ad506f1f2dc73b5100e4f3c
+```
+
+## Commands
+
+### `pay_invoice`
+
+Description: Requests payment of an invoice.
+
+Request:
+```jsonc
+{
+    "invoice": "lnbc50n1..." // BOLT11 invoice, string
+}
+```
+
+Response:
+```jsonc
+{
+    "preimage": "0123456789abcdef..." // preimage after payment, string
+}
+```
+
+### `balance`
+
+Description: Requests the balance of the wallet.
+
+Request: an empty JSON object.
+
+Response:
+```jsonc
+{
+    "balance": 100000 // balance in msat, int
```

Comment:
In typed languages, you have to inform the object you want to map your JSON into before calling the parser. In this proposal, there is no way to know if the JSON is going to be a balance or a result beforehand, which makes it impossible to *correctly* use in typed languages. Of course, typed languages can always navigate the JSON map manually, but it's not a great implementation.